trin-core/src/types/receipts.rs

```diff
@@ -0,0 +1,523 @@
+
+    pub fn rlp_append(&self, s: &mut RlpStream) {
+        match self.outcome {
+            TransactionOutcome::Unknown => {
+                s.begin_list(3);
+            }
+            TransactionOutcome::StateRoot(ref root) => {
+                s.begin_list(4);
+                s.append(root);
+            }
+            TransactionOutcome::StatusCode(ref status_code) => {
+                s.begin_list(4);
+                s.append(status_code);
+            }
+        }
+        s.append(&self.cumulative_gas_used);
+        s.append(&self.log_bloom);
+        s.append_list(&self.logs);
+    }
+}
+
+#[derive(Eq, Hash, Debug, Copy, Clone, PartialEq)]
+#[repr(u8)]
+pub enum TypedTxId {
+    EIP1559Transaction = 0x02,
+    AccessList = 0x01,
+    Legacy = 0x00,
+}
+
+impl TypedTxId {
+    pub fn try_from_wire_byte(n: u8) -> anyhow::Result<Self> {
+        match n {
+            x if x == TypedTxId::EIP1559Transaction as u8 => Ok(Self::EIP1559Transaction),
+            x if x == TypedTxId::AccessList as u8 => Ok(Self::AccessList),
+            x if (x & 0x80) != 0x00 => Ok(Self::Legacy),
+            _ => Err(anyhow!("Invalid byte selector for tx type.")),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TypedReceipt {
+    Legacy(LegacyReceipt),
+    AccessList(LegacyReceipt),
+    EIP1559Transaction(LegacyReceipt),
+}
+
+impl TypedReceipt {
+    /// Create a new receipt.
+    pub fn new(type_id: TypedTxId, legacy_receipt: LegacyReceipt) -> Self {
+        //curently we are using same receipt for both legacy and typed transaction
+        match type_id {
+            TypedTxId::EIP1559Transaction => Self::EIP1559Transaction(legacy_receipt),
+            TypedTxId::AccessList => Self::AccessList(legacy_receipt),
+            TypedTxId::Legacy => Self::Legacy(legacy_receipt),
+        }
+    }
+
+    pub fn tx_type(&self) -> TypedTxId {
+        match self {
+            Self::Legacy(_) => TypedTxId::Legacy,
+            Self::AccessList(_) => TypedTxId::AccessList,
+            Self::EIP1559Transaction(_) => TypedTxId::EIP1559Transaction,
+        }
+    }
+
+    pub fn receipt(&self) -> &LegacyReceipt {
+        match self {
+            Self::Legacy(receipt) => receipt,
+            Self::AccessList(receipt) => receipt,
+            Self::EIP1559Transaction(receipt) => receipt,
+        }
+    }
+
+    pub fn receipt_mut(&mut self) -> &mut LegacyReceipt {
+        match self {
+            Self::Legacy(receipt) => receipt,
+            Self::AccessList(receipt) => receipt,
+            Self::EIP1559Transaction(receipt) => receipt,
+        }
+    }
+
+    fn decode(tx: &[u8]) -> Result<Self, DecoderError> {
+        if tx.is_empty() {
+            // at least one byte needs to be present
+            return Err(DecoderError::RlpIncorrectListLen);
+        }
+        let id = TypedTxId::try_from_wire_byte(tx[0]);
+        if id.is_err() {
+            return Err(DecoderError::Custom("Unknown transaction"));
+        }
```

Comment:
I think the underlying error is clearer here, no reason to wrap it. Plus then it lets you:
```suggestion
        let id = TypedTxId::try_from_wire_byte(tx[0])?;
```

Reply:
We can also `map_err` here before `?`.